src/thingset_txt.cpp

```diff
@@ -8,13 +8,70 @@
 #include "thingset.h"
 #include "jsmn.h"
 
+#include <math.h>
 #include <stdbool.h>
 #include <string.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <errno.h>
 #include <cinttypes>
 
+#if !TS_PRINTF_FLOAT_SUPPORT
+
+/*
+ * Custom and simplified float to string conversion to avoid enabling float support for printf
+ *
+ * This function only supports the basic features required for JSON. Not all edge cases may have
+ * been tested. For 100% safe way use lib-c implementation.
+ */
+static int f2json(float value, int precision, char *buf, size_t size)
+{
+    int32_t integer;        // digits before the comma
+    int32_t fraction;       // digits behind the comma
+
+    if (isnan(value) || isinf(value)) {
+        /* JSON spec does not support NaN and Inf, so we need to use null instead */
+        return snprintf(buf, size, "null");
+    }
+
+    if (precision > 10) {
+        // an int32 can store max. 10 digits
+        return 0;
+    }
+
+    if (precision > 0) {
+        integer = (int32_t)value;
+
+        /* Using possibly less efficient for loop instead of pow function with large footprint to
+         * get multiplier for desired precision. Precision is increased by 1 digit to implement
+         * correct rounding prior to casting below
+         */
+        int32_t factor = 1;
+        for (int i = 0; i <= precision; i++) {
+            factor *= 10;
+        }
+
+        if (value > 0) {
+            fraction = (int32_t)((value - integer) * factor);
+        }
+        else {
+            fraction = (int32_t)((-value + integer) * factor);
+        }
+
+        /* +5 instead of +0.5 for rounding as we multiplied by 10 before */
+        fraction = (fraction + 5) / 10;
+
+        return snprintf(buf, size, "%" PRIi32 ".%.*" PRIi32,
```

Comment:
This bothers me a bit because the precision argument on integer formatting can be a bit odd (the .* specifier). It can lead to left padding with zeroes (which will throw off the meaning of the fraction); or right padding with zeroes (which I don't understand why it causes some catastrophic overflows / ronundings in our case).

Try this corner case for instance (not actually tested via unity as I'm not sure how to quickly run the suite, but I've done it through a ghetto test script):

```
char buffer [100];
float expected = 12345.54;
f2json(expected, 9, buffer, sizeof(buffer));
float actual = stof(buffer, 0);
float delta = 0.15;
TEST_ASSERT_FLOAT_WITHIN(delta, expected, actual); 
```

This tests should pass if the precision on the integer is dropped.

Reply:
That's an interesting edge case. I got `12345.076149043` as a result. If I reduce the precision from 9 to 8, it returns `12345.54003904`, which is the same as with the original printf function. The issue, I guess, is that the significant digits of a float32 is maximum 9. So we should probably consider that as the limitation instead of just the digits behind the comma. I'll think about how this could be improved without too much overhead.

Reply:
Mhh, doesn't seem to be as easy as it seems... found more edge cases for rounding.

Reply:
I would leave it as it is and maybe reduce the acceptable precision to a max of 6 (and note that in the docs). Also as I mentioned, removing `.*` in the format specifier seems to work much better.

Reply:
You are right - I was being silly, trying to break it in some improbable ways (calling the function with a given precision then messing with the precision param of the snprintf and getting rubbish out). You actually have a test case for the situation you expose above (and which I haven't thought of) so it's all good.